File: tp03/TP03Q13 - Fila Flexível em Java/Serie.java
/* BIBLIOTECA*/
import java.io.*;

public class Serie{
	final static int tamPubIn = 100;
	final static int tamHtml = 100;
	final static int tamLinha = 10000;
	static int tam1 = 0;
	static int tam2 = 0;

	/* ATRIBUTOS */
	private String nome;
  	private String formato;
  	private String duracao;
  	private String paisDeOrigem;
  	private String idiomaOriginal;
 	private String emissoraOriginal;
  	private String transmissaoOriginal;
  	private int numTemporadas;
  	private int numEpisodios;

  	/* CONSTRUTOR PADRAO(PRIMARIO) */
  	public Serie(){
		this.nome = "";
		this.formato = "";
		this.duracao = "";
		this.paisDeOrigem = "";
		this.idiomaOriginal = "";
		this.emissoraOriginal = "";
		this.transmissaoOriginal = "";
		this.numTemporadas = 0;
		this.numEpisodios = 0;
  }

  /* CONSTRUTOR COM PARAMETROS(SECUNDARIO) */
  public Serie(String nome, String formato, String duracao, String paisDeOrigem, String idiomaOriginal, String emissoraOriginal, String transmissaoOriginal, int numTemporadas, int numEpisodios){
		super();
		this.nome = nome;
		this.formato = formato;
		this.duracao = duracao;
		this.paisDeOrigem = paisDeOrigem;
		this.idiomaOriginal = idiomaOriginal;
		this.emissoraOriginal = emissoraOriginal;
		this.transmissaoOriginal = transmissaoOriginal;
		this.numTemporadas = numTemporadas;
		this.numEpisodios = numEpisodios;
	}

  	/* GETS E SETS */
  	public String getNome(){
		return nome;
	}
	public void setNome(String nome){
		this.nome = nome;
	}
	public String getFormato(){
		return formato;
	}
	public void setFormato(String formato){
		this.formato = formato;
	}
	public String getDuracao(){
		return duracao;
	}
	public void setDuracao(String duracao){
		this.duracao = duracao;
	}
	public String getPaisDeOrigem(){
		return paisDeOrigem;
	}
	public void setPaisDeOrigem(String paisDeOrigem){
		this.paisDeOrigem = paisDeOrigem;
	}
	public String getIdiomaOriginal(){
		return idiomaOriginal;
	}
	public void setIdiomaOriginal(String idiomaOriginal){
		this.idiomaOriginal = idiomaOriginal;
	}
	public String getEmissoraOriginal(){
		return emissoraOriginal;
	}
	public void setEmissoraOriginal(String emissoraOriginal){
		this.emissoraOriginal = emissoraOriginal;
	}
	public String getTransmissaoOriginal(){
		return transmissaoOriginal;
	}
	public void setTransmissaoOriginal(String transmissaoOriginal){
		this.transmissaoOriginal = transmissaoOriginal;
	}
	public int getNumTemporadas(){
		return numTemporadas;
	}
	public void setNumTemporadas(int numTemporadas){
		this.numTemporadas = numTemporadas;
	}
	public int getNumEpisodios(){
		return numEpisodios;
	}
	public void setNumEpisodios(int numEpisodios){
		this.numEpisodios = numEpisodios;
	}

	/* METODO CLONE */
	public Serie cloneSerie(){
		Serie cloneSerie = new Serie();
		cloneSerie.nome = nome;
		cloneSerie.formato = formato;
		cloneSerie.duracao = duracao;
		cloneSerie.paisDeOrigem = paisDeOrigem;
		cloneSerie.idiomaOriginal = idiomaOriginal;
		cloneSerie.emissoraOriginal = emissoraOriginal;
		cloneSerie.transmissaoOriginal = transmissaoOriginal;
		cloneSerie.numTemporadas = numTemporadas;
		cloneSerie.numEpisodios = numEpisodios;
		return cloneSerie;
	}

	/* METODO PARA CONFERIR SE CHEGOU AO FIM DO ARQUIVO */
	public static boolean conferirFim(String linha){ /* CONDI????????????O DE PARADA */
		boolean resp;
		if(linha.length()==3 && linha.charAt(0)=='F' && linha.charAt(1)=='I' && linha.charAt(2)=='M'){
			resp = true;
		}else{
			resp = false;
		}
		return resp;
	}
	
	/* METODO PARA IMPRIMIR VALORES NA ORDEM CORRRETA SEGUNDO O PUB.OUT */
	public Serie[] imprimir(){
		Serie[] S = new Serie[tamPubIn];
		MyIO.println(nome + " " + formato + " " + duracao + " " + paisDeOrigem + " " +
		idiomaOriginal + " " + emissoraOriginal + " " + transmissaoOriginal + " " + numTemporadas + " " + numEpisodios );

		return S;
	}
	
	/* M??TODO PARA VERIFICAR SE ?? UMA LETRA OU NUMERO */
	public static boolean isLetraOrNumber(char a) {
    	return (a >= '!' || a >= 'a' && a <= 'z' || a >= 'A' && a <= 'Z' || a >= '0' && a <= '9' || a == ')' || a == '(' || a == '-' || a == ';');
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO NOME */
  	public static String encontrarNome (String[] linha){
  		String nome = "";
  		for(int i = 2; i < linha.length; i = i+2){
  			if(!linha[i].equals("") && isLetraOrNumber(linha[i].charAt(0))){
  				nome = nome + "" + linha[i] + "";	
  			}
  		}
  		return nome;
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO FORMATO */
  	public static String encontrarFormato (String[] linha){
  		String formato = "";
  		for(int i = 2; i < linha.length; i = i+2){
  			if(!linha[i].equals("") && isLetraOrNumber(linha[i].charAt(0))){
  				formato = formato + "" + linha[i] + "";	
  			}
  		}
  		return formato;
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO DURACAO */
  	public static String encontrarDuracao (String[] linha){
  		String duracao = "";
  		for(int i = 2; i < linha.length; i = i+2){
  			if(!linha[i].equals("") && isLetraOrNumber(linha[i].charAt(0))){
  				duracao = duracao + "" + linha[i] + "";	
  			}
  		}
  		return duracao;
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO PAIS DE ORIGEM */
  	public static String encontrarPaisDeOrigem (String[] linha){
  		String paisDeOrigem = "";
  		for(int i = 2; i < linha.length; i = i+2){
  			if((!linha[i].equals("") && isLetraOrNumber(linha[i].charAt(0))) || (linha[i].length() >= 1 && linha[i].charAt(0) == ' ')){
  				if(!linha[i].equals("&#160;") && !linha[i].equals("&nbsp;")){
  					paisDeOrigem = paisDeOrigem + "" + linha[i] + "";	
  				}
  			} 
  		}
  		return paisDeOrigem;
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO IDIOMA ORIGINAL */
  	public static String encontrarIdiomaOriginal (String[] linha){
  		String idiomaOriginal = "";
  		for(int i = 2; i < linha.length; i = i+2){
  			if((!linha[i].equals("") && isLetraOrNumber(linha[i].charAt(0))) || (linha[i].length() > 4 && linha[i].charAt(0) == ' ')){
  				idiomaOriginal = idiomaOriginal + "" + linha[i] + "";	
  			}
  		}
  		return idiomaOriginal;
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO EMISSORA ORIGINAL */
  	public static String encontrarEmissora (String[] linha){
  		String emissora = "";
  		for(int i = 2; i < linha.length; i = i+2){
  			if(!linha[i].equals("") && !linha[i].equals(" ") && isLetraOrNumber(linha[i].charAt(0))){
  				emissora = emissora + "" + linha[i] + "";
  			} else if(linha.length > 6 && linha[7].equals("br /")){
  				emissora = emissora + "" + linha[i] + "";
  			} else if(linha.length > 12 && linha[13].equals("br /") && !linha[i].equals(" ")){
  				emissora = emissora + "" + linha[i] + "";
  			}
  		}
  		return emissora;
  	}
  	
  	/* METODO PARA ENCONTRAR O ATRIBUTO TRANSMISSAO ORIGINAL */
  	public static String encontrarTransmissaoOriginal (String[] linha){
  		String transmissao = "";
  		for(int i = 2; i < linha.length; i = i+2){
			linha[i] = linha[i].replace("&#160;(","(");
			linha[i] = linha[i].replace("&nbsp;(","(");
			linha[i] = linha[i].replace(" ??? ","!");
			linha[i] = linha[i].replace(" - ",";");
			if(linha[i].equals(" de ") || (!linha[i].equals("") && isLetraOrNumber(linha[i].charAt(0)))){
				linha[i] = linha[i].replace("!"," ??? ");
				linha[i] = linha[i].replace(";"," - ");
				transmissao = transmissao + "" + linha[i] + "";	
			}
  		}
  		return transmissao;
  	}
   	
	/* METODO NECESSERIO PARA FAZER A LEITURA DO HTML */
  	public static Serie leituraHtml(String linhasPubIn) throws IOException {
    	Serie s = new Serie();
    	String linha = new String();
    	String[] aux2 = new String[100];
    	String aux1 = new String();
  		String aux3 = new String();  
		Arq.openRead("/tmp/series/" + linhasPubIn, "UTF-8");
    	FileReader fileReader = new FileReader("/tmp/series/" + linhasPubIn);
		BufferedReader bufferedReader = new BufferedReader(fileReader);
    	while((linha = bufferedReader.readLine()) != null) {
			if(linha.contains("firstHeading") == true){
				aux1 = linha.replace(">","<");
				aux2 = aux1.split("<");
				s.nome = encontrarNome(aux2);
			} else if(linha.contains("<tr>") == true || linha.contains("<tbody><tr>") == true){ 
				linha = bufferedReader.readLine(); 
				if(linha.contains("Formato</td>") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux2 = aux1.split("<");
					s.formato = encontrarFormato(aux2);
				} else if(linha.contains("Dura????o") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux2 = aux1.split("<");
					s.duracao = encontrarDuracao(aux2);
				} else if(linha.contains("Pa??s de origem") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux2 = aux1.split("<");
					s.paisDeOrigem = encontrarPaisDeOrigem(aux2);
				} else if(linha.contains("Idioma original") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux2 = aux1.split("<");
					s.idiomaOriginal = encontrarIdiomaOriginal(aux2);
				} else if(linha.contains("Emissora de televis??o original") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux2 = aux1.split("<");
					s.emissoraOriginal = encontrarEmissora(aux2);
				} else if(linha.contains("Transmiss??o original") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux2 = aux1.split("<");
					s.transmissaoOriginal = encontrarTransmissaoOriginal(aux2);
				} else if(linha.contains("N.?? de temporadas") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux3 = aux1.replace(" ", "<");
					aux2 = aux3.split("<");
					s.numTemporadas = Integer.parseInt(aux2[6]);
				} else if(linha.contains("N.?? de epis??dios") == true){
					linha = bufferedReader.readLine(); 
					aux1 = linha.replace(">","<");
					aux3 = aux1.replace(" ", "<");
					aux2 = aux3.split("<");
					s.numEpisodios = Integer.parseInt(aux2[6]);
				} 
			}
		}
    	//s.imprimir();  
    	bufferedReader.close(); 
    	Arq.close();
    	return s;
	}
	
	/* METODO NECESSARIO PARA FAZER A LEITURA DO PUB.IN */
  	public static String[] leituraPubIn(int i){
    	String[] linhas = new String[tamPubIn];
    	if(i == 1){ 
      		linhas[tam1] = MyIO.readLine();
      		while(conferirFim(linhas[tam1]) == false){ /* LEITURA DO ARQUIVO ENQUANTO ESSE N??????O CHEGAR NO FIM */
				tam1++;
     	 		linhas[tam1] = MyIO.readLine();
			}
		} else {
			tam2 = MyIO.readInt();
			int j = 0;
        	while(j < tam2){ /* LEITURA DO ARQUIVO ENQUANTO ESSE N??????O CHEGAR NO FIM */
          		linhas[j] = MyIO.readLine();
			 	j++;
			}
		}
		return linhas;
	}
	
	/* MAIN DO CODIGO QUE CHAMA OS METODOS */
	public static void main(String[] args)throws Exception{
		MyIO.setCharset("UTF-8");
		String[] linhasPubIn = leituraPubIn(1);
		String[] linhasPilha = leituraPubIn(2); 
		Serie[] linhasRemovidas = new Serie[100];
		Serie[] serie = new Serie[tam1];
		int i = 0, j = 0;
		try {
			while(i < tam1){
				serie[i] = leituraHtml(linhasPubIn[i]);
				i++;
			}
		}catch(IOException e) {
  			e.printStackTrace();
		}
    	Fila filaSeries = new Fila();
		for(i = 0; i < tam1; i++){
			filaSeries.inserir(serie[i]);	
		}
		i = 0;
		String[] aux = new String[5];
		while(i < tam2){
			aux = linhasPilha[i].split(" ");
			if(aux[0].compareTo("I") == 0){
				filaSeries.inserir(leituraHtml(aux[1]));
			} else if(aux[0].compareTo("R") == 0){
				linhasRemovidas[j] = filaSeries.remover();
				j++;
			} 
			i++;	
		}
		/* for(i = 0; i < j; i++){
			MyIO.println("(R) " + linhasRemovidas[i].getNome());
		}
		filaSeries.mostrar(); */
	}
}

class Fila {
	private Celula primeiro;
	private Celula ultimo;
	private int aux = 0;

	public Fila() {
		primeiro = new Celula();
		ultimo = primeiro;
	}

	public void inserir(Serie x) throws Exception {
		if(aux == 5){
			remover();
		}

		ultimo.prox = new Celula(x);
		ultimo = ultimo.prox;
		aux++;

		int count = 0;
		Celula tmp = primeiro.prox;
		while(tmp != null){
			count = count + tmp.elemento.getNumTemporadas();
			tmp = tmp.prox;
		}

		count = (int) Math.round((double) count/aux) ;
		MyIO.println(count);
	}

    public Serie remover() throws Exception {
		if (primeiro == ultimo) {
			throw new Exception("Erro ao remover!");
		}

        Celula tmp = primeiro;
		primeiro = primeiro.prox;
		Serie resp = primeiro.elemento;
        tmp.prox = null;
        tmp = null;
		aux--;

		return resp;
	}

    public void mostrar() {
		for(Celula i = primeiro.prox; i != null; i = i.prox) {
			i.elemento.imprimir();
		}
	}
}

class Celula {
	public Serie elemento; // Elemento inserido na celula.
	public Celula prox; // Aponta a celula prox.

    public Celula(){
		this.prox = null;
    }

    //Construtor da classe
	public Celula(Serie elemento) {
        this.elemento = elemento;
        this.prox = null;
	}
}

File: tp02/TP02Q08 - Lista Sequencial em C/Serie.c
// BIBLIOTECAS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

//ESTRUTURA
typedef struct {
    char* nome;
    char* formato;
    char* duracao;
    char* paisOriginal;
    char* idiomaOriginal;
    char* emissoraOriginal;
    char* transmissaoOriginal;
    int numTemporadas;
    int numEpisodios;
}Series;

int tamLista = 0;
//METODOS DA LISTA

void inserirInicio(Series* x, Series** array) {
   int i;
   //validar insercao
   if(tamLista >= 100){
      printf("Erro ao inserir Inicio!");
      exit(1);
   } 
   //levar elementos para o fim do array
   for(i = tamLista; i > 0; i--){
      array[i] = array[i-1];
   }
   tamLista++;
   array[0] = x;
}

void inserirFim(Series* x, Series** array) {
   //validar insercao
   if(tamLista >= 100){
      printf("Erro ao inserir Fim!");
      exit(1);
   }
   array[tamLista] = x;
   tamLista++;
}

void inserir(Series* x, int pos, Series** array) {
   int i;
   //validar insercao
   if(tamLista >= 100 || pos < 0 || pos > tamLista){
      printf("Erro ao inserir!");
      exit(1);
   }
   //levar elementos para o fim do array
   for(i = tamLista; i > pos; i--){
      array[i] = array[i-1];
   }
   array[pos] = x;
   tamLista++;
}

Series* removerInicio(Series** array) {
   int i = 0;
   Series* resp = (Series *)calloc(1000, sizeof(Series));
   //validar remocao
   if (tamLista == 0) {
      printf("Erro ao remover!");
      exit(1);
   }
   resp = array[0];
   tamLista--;
   for(i = 0; i < tamLista; i++){
      array[i] = array[i+1];
   }
   return resp;
}

Series* removerFim(Series** array) {
   //validar remocao
   if (tamLista == 0) {
      printf("Erro ao remover!");
      exit(1);
   }
   return array[--tamLista];
}

Series* remover(int pos, Series** array) {
   int i = 0;
   Series* resp = (Series *)calloc(1000, sizeof(Series));
   //validar remocao
   if (tamLista == 0 || pos < 0 || pos >= tamLista) {
      printf("Erro ao remover!");
      exit(1);
   }
   resp = array[pos];
   tamLista--;
   for(i = pos; i < tamLista; i++){
        array[i] = array[i+1];
   }
   return resp;
}

//METODOS DA SERIE

//PROCEDIMENTOS PARA SETAR A ESTRUTURA
void setNome(char* atributo, Series* serie){
    serie->nome = atributo; 
}
void setFormato(char* atributo, Series* serie){
    serie->formato = atributo; 
}
void setDuracao(char* atributo, Series* serie){
    serie->duracao = atributo; 
}
void setPaisOriginal(char* atributo, Series* serie){
    serie->paisOriginal = atributo; 
}
void setIdiomaOriginal(char* atributo, Series* serie){
    serie->idiomaOriginal = atributo; 
}
void setEmissoraOriginal(char* atributo, Series* serie){
    serie->emissoraOriginal = atributo; 
}
void setTransmissaoOriginal(char* atributo, Series* serie){
    serie->transmissaoOriginal = atributo; 
}
void setNumTemporadas(int atributo, Series* serie){
    serie->numTemporadas = atributo; 
}
void setNumEpisodios(int atributo, Series* serie){
    serie->numEpisodios = atributo; 
}

//FUNCÕES PARA FAZER O GET DA ESTRUTURA
char* getNome(Series* serie){
    return serie->nome;
}
char* getFormato(Series* serie){
    return serie->formato;
}
char* getDuracao(Series* serie){
    return serie->duracao;
}
char* getPaisOriginal(Series* serie){
    return serie->paisOriginal;
}
char* getIdiomaOriginal(Series* serie){
    return serie->idiomaOriginal;
}
char* getEmissoraOriginal(Series* serie){
    return serie->emissoraOriginal ;
}
char* getTransmissoraOriginal(Series* serie){
    return serie->transmissaoOriginal;
}
int getNumTemporadas(Series* serie){
    return serie->numTemporadas;
}
int getNumEpisodios(Series* serie){
    return serie->numEpisodios;    
}

//FUNCAO QUE CONFERE SE CHEGOU AO FIM DO ARQUIVO
bool isFim(char* linha) {
    return (linha[0] == 'F' && linha[1] == 'I' && linha[2] == 'M');
}

//FUNCAO QUE RETORNA SE O CARACTER É UMA LETRA OU UM NUMERO
bool isLetraOrNumber(char* a){
	return (*a >= 'a' && *a <= 'z' || *a >= 'A' && *a <= 'Z' || *a >= '0' && *a <= '9' || *a == ' ' || *a == '(' || *a == ')');
}

//FUNCAO QUE RETORNA SE O CARACTER É UMA LETRA OU UM NUMERO
bool isLetraOrNumber2(char* a){
	return (*a >= 'a' && *a <= 'z' || *a >= 'A' && *a <= 'Z' || *a >= '0' && *a <= '9');
}

//FUNCAO QUE RETORNA SE O CARACTER É UM NUMERO
bool isNumber(char* a){
	return (*a >= '0' && *a <= '9');
}

//FUNCAO DE IMPRIMIR OS ATRIBUTOS NA ORDEM DO PUB.IN
void imprimir(Series* s){
	printf("%s %s %s %s %s %s %s %d %d \n", getNome(s), getFormato(s), getDuracao(s), getPaisOriginal(s), getIdiomaOriginal(s), getEmissoraOriginal(s), getTransmissoraOriginal(s), getNumTemporadas(s), getNumEpisodios(s));
}

//FUNCAO QUE TRATA O ATRIBUTO
char* pegarAtributo(char* linha, int opc){
	int indexAtributo = 0, indexColuna = 0;
	char* atributo = (char*)calloc(100, sizeof(char));
	while(indexColuna < strlen(linha)-1){
		if(linha[indexColuna] == '>'){
			indexColuna++;
			switch (opc){
    			case 1:
					if(isLetraOrNumber2(&linha[indexColuna])){
						while(linha[indexColuna] != '<'){
							atributo[indexAtributo++] = linha[indexColuna++];
						}
					}
					break;
			
				case 2:
					if(isLetraOrNumber(&linha[indexColuna])){
						while(linha[indexColuna] != '<'){
							atributo[indexAtributo++] = linha[indexColuna++];
						}
					} 
					break;
						
				case 3:
					if(isNumber(&linha[indexColuna])){
						while(linha[indexColuna] != '<'){
							atributo[indexAtributo++] = linha[indexColuna++];
						}
					}
					break;
					
				case 4:
					if(isLetraOrNumber(&linha[indexColuna])){
						while(linha[indexColuna] != '<'){
							atributo[indexAtributo++] = linha[indexColuna++];
						}
					} else {
						if(linha[indexColuna++] == '&'){
							atributo[indexAtributo++] = '(';
						}
					}	
					break;
			}			
		}	
		indexColuna++;
	}
	return atributo;
}

//FUNCAO QUE TRATA NOME
char* tratarNome(char** html){
    char *nome = (char *)calloc(50, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "firstHeading") == NULL)indexLinha++;
    nome = pegarAtributo(&html[indexLinha][indexColuna], 1);
    return nome;
}

//FUNCAO QUE TRATA FORMATO
char* tratarFormato(char** html){
    char *formato = (char *)calloc(100, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "Formato") == NULL)indexLinha++;
    indexLinha++;
    formato = pegarAtributo(&html[indexLinha][indexColuna], 2);
    return formato;
}

//FUNCAO QUE TRATA DURACAO
char* tratarDuracao(char** html){
    char *duracao = (char *)calloc(100, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "Duração") == NULL)indexLinha++;
    indexLinha++;
    duracao = pegarAtributo(&html[indexLinha][indexColuna], 2);
    return duracao;
}

//FUNCAO QUE TRATA PAIS ORIGINAL
char* tratarPaisOriginal(char** html){
    char *paisOriginal = (char *)calloc(100, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "País de origem") == NULL)indexLinha++;
    indexLinha++;
    paisOriginal = pegarAtributo(&html[indexLinha][indexColuna], 2);
    return paisOriginal;
}

//FUNCAO QUE TRATA IDIOMA ORIGINAL
char* tratarIdiomaOriginal(char** html){
    char *idiomaOriginal = (char *)calloc(100, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "Idioma original") == NULL)indexLinha++;
    indexLinha++;
    idiomaOriginal = pegarAtributo(&html[indexLinha][indexColuna], 2);
    return idiomaOriginal;
}

//FUNCAO QUE TRATA EMISSORA ORIGINAL
char* tratarEmissoraOriginal(char** html){
    char *emissoraOriginal = (char *)calloc(100, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "Emissora de televisão original") == NULL)indexLinha++;
    indexLinha++;
    emissoraOriginal = pegarAtributo(&html[indexLinha][indexColuna], 2);
    return emissoraOriginal;
}

//FUNCAO QUE TRATA EMISSORA ORIGINAL
char *tratarTransmissoraOriginal(char **html)
{
    char *transmissoraOriginal = (char *)calloc(100, sizeof(char));
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "Transmissão original") == NULL)indexLinha++;
    indexLinha++;
    transmissoraOriginal = pegarAtributo(&html[indexLinha][indexColuna], 4);
    return transmissoraOriginal;
}

//FUNCAO QUE TRATA NUMERO DE TEMPORADAS
int tratarNumTemporadas(char** html){
    int numTemporadas = 0;
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "N.º de temporadas") == NULL)indexLinha++;
    indexLinha++;
    numTemporadas = atoi(pegarAtributo(&html[indexLinha][indexColuna], 3));
    return numTemporadas;
}

//FUNCAO QUE TRATA NUMERO DE EPISODIOS
int tratarNumEpisodios(char** html){
    int numEpisodios = 0;
    int indexLinha = 0, indexColuna = 0, indexAtributo = 0;
    while(strstr(html[indexLinha], "N.º de episódios") == NULL)indexLinha++;
    indexLinha++;
    numEpisodios = atoi(pegarAtributo(&html[indexLinha][indexColuna], 3));
    return numEpisodios;
}

//FUNCAO DE LEITURA DOS ARQUIVOS HTML
char** lerHtml(char* caminho) {
    char** html = (char**)calloc(3000,sizeof(char*));
    int i = 0;
    for (i = 0; i < 3000; i++) {
        html[i] = (char*)calloc(3000,sizeof(char));
    }
    FILE* arquivo = fopen(caminho, "r");
    if (arquivo != NULL) {
        i = 0;
        while (!feof(arquivo)) {
            fgets(html[i],3000,arquivo);
            i++;
        }
    } else {
        printf("ERRO AO ABRIR ARQUIVO \n");
    }
    fclose(arquivo);
    return html;
}

//FUNCAO QUE CHAMA O METODO DE SETAR A ESTRUTURA
Series* ler(char* caminho, Series* serie){
    char** html = lerHtml(caminho);
    setNome(tratarNome(html), serie);
    setFormato(tratarFormato(html), serie);
    setDuracao(tratarDuracao(html), serie);
    setPaisOriginal(tratarPaisOriginal(html), serie);
    setIdiomaOriginal(tratarIdiomaOriginal(html), serie);
    setEmissoraOriginal(tratarEmissoraOriginal(html), serie);
    setTransmissaoOriginal(tratarTransmissoraOriginal(html), serie);
    setNumTemporadas(tratarNumTemporadas(html), serie);
    setNumEpisodios(tratarNumEpisodios(html), serie); 

    return serie;
}

Series* pegarSerie(char* aux){
    Series* serie = (Series *)calloc(1000, sizeof(Series));
    char* caminho = (char*)calloc(15+strlen(aux), sizeof(char));
    strcpy(caminho,"/tmp/series/"); 
    strcat(caminho, aux);
    caminho[strlen(caminho)-1] = '\0';
    serie = ler(caminho, serie);

    return serie;
}

void inserirRemover(char** str, Series** array, int quant){
    Series* serieLista = (Series *)calloc(1, sizeof(Series));
    Series* serieRemovida = (Series *)calloc(1, sizeof(Series));
    int i = 0;
    while(i < quant){
        char aux[100] = "";
        char pos[100] = "";
        int k = 0, posicao = 0, j = 0;
        if(str[i][0] == 'I' && str[i][1] == 'I'){
            j = 0;
            for(k = 3; k < strlen(str[i]); k++){
                aux[j] = str[i][k];
                j++;
            }
            serieLista = pegarSerie(aux);
            inserirInicio(serieLista, array);
        } else if(str[i][0] == 'I' && str[i][1] == 'F'){
            j = 0;
            for(k = 3; k < strlen(str[i]); k++){
                aux[j] = str[i][k];
                j++;
            }
            serieLista = pegarSerie(aux);
            inserirFim(serieLista, array);
        } else if(str[i][0] == 'I' && str[i][1] == '*'){
            k = 3;
            j = 0;
            while(str[i][k] != ' '){
                pos[j] = str[i][k];
                k++; j++;
            }
            posicao = atoi(pos);
            int tam = 0;
            if(str[i][5] == ' '){
                tam = 6;
            } else {
                tam = 5;
            }
            j = 0;
            for(k = tam; k < strlen(str[i]); k++){
                aux[j] = str[i][k];
                j++;
            }
            serieLista = pegarSerie(aux);
            inserir(serieLista, posicao, array);
        } else if(str[i][0] == 'R' && str[i][1] == 'I'){
            serieRemovida = removerInicio(array);
            printf("(R) %s \n", getNome(serieRemovida));
        } else if(str[i][0] == 'R' && str[i][1] == 'F'){  
            serieRemovida = removerFim(array);
            printf("(R) %s \n", getNome(serieRemovida));
        } else if(str[i][0] == 'R' && str[i][1] == '*'){  
            k = 3; 
            j = 0;
            while(str[i][k] != '\0'){
                pos[j] = str[i][k];
                k++; j++;
            }
            posicao = atoi(pos); 
            serieRemovida = remover(posicao, array);
            printf("(R) %s \n", getNome(serieRemovida));
        }
    i++;
    }
}

//MAIN DO CODIGO
int main() {
	Series** series = (Series **)calloc(1000, sizeof(Series *));
    Series** array = (Series **)calloc(1000, sizeof(Series *));
	char** caminho = (char**)calloc(100, sizeof(char*));
    char** entradas = (char **)calloc(100 , sizeof(char*));
    char** entradas2 = (char **)calloc(100 , sizeof(char*));
    int i, k = 0, quant = 0;
    char* aux = (char *)calloc(100 , sizeof(char));
    for (i = 0; i < 100; i++) {
        entradas[i] = (char*)calloc(2000 , sizeof(char));
        entradas2[i] = (char*)calloc(2000 , sizeof(char));
        series[i] = (Series*)calloc(100, sizeof(Series));
        array[i] = (Series*)calloc(100, sizeof(Series));
    } 
    i = 0;
    fgets(entradas[i], 1000,stdin);
    while (!isFim(entradas[i])) {
        i++; k++;
        fgets(entradas[i], 1000,stdin);
    }
    i = 0;
    quant = atoi(fgets(aux, 10,stdin));
    while (i < quant) {
        fgets(entradas2[i], 1000,stdin);
        i++; 
    }
    i = 0;
    while(i < k){
    	caminho[i] = (char*)calloc(15+strlen(entradas[i]), sizeof(char));
    	strcpy(caminho[i],"/tmp/series/"); 
        strcat(caminho[i], entradas[i]);
        caminho[i][strlen(caminho[i])-1] = '\0';
        ler(caminho[i], series[i]);
        inserirFim(series[i], array);
        i++;
    } 
    inserirRemover(entradas2, array, quant);

    for(i = 0; i < tamLista; i++){
        imprimir(array[i]);
    }

    free(series);
    free(array);
    free(caminho);
    free(entradas);
    free(entradas2);
    
    return 0;
}
